File: packages/typescript/src/boundary.ts
import { ClientError } from "./transport.js";

/** Validate supported evaluation inputs before issuing a request. */
export function validateSubmission(resource: "quotes" | "evaluations", input: unknown): void {
  const object = (value: unknown): value is Record<string, unknown> =>
    value !== null && typeof value === "object" && !Array.isArray(value);
  const allowed = new Set(resource === "quotes"
    ? ["selection", "model", "coverage", "max_charge_microusd"]
    : ["model", "eval", "budget_usd", "max_charge_microusd", "provider", "coverage", "name", "metadata", "split"]);
  const fail = () => { throw new ClientError("unsupported_feature", "Choose an admitted eval and a managed model or checked connection."); };
  if (!object(input) || Object.keys(input).some((key) => !allowed.has(key))) return fail();
  const model = input.model;
  if (typeof model === "string" && resource === "evaluations") {
    if (!model || /^(checkpoint|agent|artifact):/.test(model)) return fail();
  } else if (object(model)) {
    const keys = model.kind === "managed" ? ["kind", "route_id"] : model.kind === "connection" ? ["kind", "connection_id"] : [];
    if (keys.length === 0 || Object.keys(model).length !== keys.length || keys.some((key) => !(key in model))) return fail();
  } else return fail();
  if (resource === "quotes") {
    const selection = input.selection;
    if (!object(selection)) return fail();
    const keys = Object.keys(selection);
    if (!(keys.length === 1 && keys[0] === "profile_ids") &&
        !("eval" in selection && keys.every((key) => ["eval", "provider", "split"].includes(key)))) return fail();
  }
}

File: packages/python/src/kimpton_evalrouter/_boundary.py
"""Validate supported evaluation inputs before issuing a request."""

from ._transport import ClientError


def validate_submission(resource, body):
    allowed = (
        {"selection", "model", "coverage", "max_charge_microusd"}
        if resource == "quotes"
        else {
            "model",
            "eval",
            "budget_usd",
            "max_charge_microusd",
            "provider",
            "coverage",
            "name",
            "metadata",
            "split",
        }
    )
    if not isinstance(body, dict) or set(body) - allowed:
        raise ClientError(
            "unsupported_feature", "This submission contains unsupported input fields."
        )
    model = body.get("model")
    if isinstance(model, str) and resource == "evaluations":
        valid = bool(model) and not model.startswith(("checkpoint:", "agent:", "artifact:"))
    elif isinstance(model, dict):
        keys = {"managed": {"kind", "route_id"}, "connection": {"kind", "connection_id"}}
        kind = model.get("kind")
        valid = isinstance(kind, str) and kind in keys and set(model) == keys[kind]
    else:
        valid = False
    if resource == "quotes":
        selection = body.get("selection")
        valid = (
            valid
            and isinstance(selection, dict)
            and (
                set(selection) == {"profile_ids"}
                or "eval" in selection
                and not set(selection) - {"eval", "provider", "split"}
            )
        )
    if not valid:
        raise ClientError(
            "unsupported_feature",
            "Choose an admitted eval and a managed model or checked connection.",
        )
